### ISSUE TITLE
Log spammed with warnings about missing file

### ISSUE THREAD

**Issue by dchristl** [Team-xManager/xManager#326]
### 🧐 Have you fully read the README and viewed similar issues?

- [X] I have read the README
- [X] I have checked older issues, open and closed
- [x] I did a clean install of xManager
- [X] This is an xManager app issue -- NOT a Spotify mod issue

### ℹ Information

```markdown
- Device Model:  
Xiaomi Poco F3
- Android Version:  
12
- ROM Version:  
Lineage OS 19
- xManager Version:  
4.9
```

### 📝 Description

1. Open Logcat
2. Open XManager

You can't see any other log entries, because the log entries comes multiple times per second (see screenshot). Only a force stop of Xmanager helps to solve this issue. 

### 📸 Screenshots

![image](https://github.com/Team-xManager/xManager/assets/2836933/f646cb43-833c-4419-9e16-4175175bfa5e)

**Comment by energ0n** [comment]
@dchristl are you still experiencing this issue as of xManager v5.4?

**Comment by dchristl** [comment]
Thank you for your answer @energ0n , but no change at all. Same behaviour with a fresh install and the newest version.

**Comment by energ0n** [comment]
> no change at all. Same behaviour with a fresh install and the newest version.

Are you still running Lineage? Can you confirm that the behavior exists even when disabling Lineage Privacy guard settings (or specifically giving xManager all storage-related permissions in Privacy guard)? 

When I test this on a Lineage rom and have privacy guard enabled and blocking xmanager storage permissions, the same behavior you showed in log happens. Disabling Privacy guard and the error goes away.

**Comment by dchristl** [comment]
Yes, I'm still runing LineageOS. I'm not quite sure what you mean by privacy guard. xManager has all relevant permissions in the privacy section. If I create this file (empty) the error message changes to can not extract apk, so I think the file access is granted.

**Comment by energ0n** [comment]
> Yes, I'm still runing LineageOS. I'm not quite sure what you mean by privacy guard. xManager has all relevant permissions in the privacy section
> 
> If I create this file (empty) the error message changes to can not extract apk, so I think the file access is granted.

Hmm, I see. LineageOS used to have an additional "privacy guard" submenu with more granular control but it seems this went away around Lineage 17.  Since you originally mentioned Lineage v19+ so I suppose you don't have that now.. although your original log seems to be behaving as if it were.. strange.

On a fresh app install (or force close + wipe app data), on first run you should be asked 3 things:
- allow notifications (this can be allowed or denied with no consequence)
- storage permission (box pops up with a setup button, click setup.. popup asking to allow storage perm -> "Allow")
- installation permission (box pops up with a setup button, click setup.. new page loads.. "install unknown apps" with toggle to "allow from this source" -> toggle to on position)

Storage and installation perms are both allowed/toggled on? 

- In xManager, there's a beaker icon in the top right corner, can you share the APK location set in that menu?
- Can you share what your log spam looks like while running xManager v5.4?
- Do you have to download an APK for the spam to start happening? If not, does something specific trigger this?
  - if so, does this happen when using either the "mirror link" option or "download" option (assuming you are running A13 or earlier.. A14 only has a "Download" button and has users download from MediaFire)?
  - if so, can you confirm with a File manager, adb or whatever that a file exists (or doesn't exist) at the path(s) being spammed in the log?

**Comment by dchristl** [comment]
I've also reached Lineage 20 in the meantime, if that makes a difference.

> On a fresh app install (or force close + wipe app data), on first run you should be asked 3 things:
allow notifications (this can be allowed or denied with no consequence)
storage permission (box pops up with a setup button, click setup.. popup asking to allow storage perm -> "Allow")
installation permission (box pops up with a setup button, click setup.. new page loads.. "install unknown apps" with toggle to "allow from this source" -> toggle to on position)

Yes, all of that was asked, and I confirmed it.

> Storage and installation perms are both allowed/toggled on?

Yes, because download works.

> In xManager, there's a beaker icon in the top right corner, can you share the APK location set in that menu?

![Screenshot_20240413-080301_xManager](https://github.com/Team-xManager/xManager/assets/2836933/6f65db47-0f40-4346-83ec-f64db734b5d4)

> Can you share what your log spam looks like while running xManager v5.4?

It's always the same output over and over again and stops only on force closing the app. 

```
04-13 08:05:09.429  5825  5825 W ziparchive: Unable to open '/storage/emulated/0/xManager/Patched Lite (xManager).apk': No such file or directory
04-13 08:05:09.429  5825  5825 E 3fff0e.xmanager: Failed to open APK '/storage/emulated/0/xManager/Patched Lite (xManager).apk': I/O error
04-13 08:05:09.429  5825  5825 W PackageManager: Failure to parse package archive
04-13 08:05:09.429  5825  5825 W PackageManager: android.content.pm.PackageParser$PackageParserException: Failed to parse /storage/emulated/0/xManager/Patched Lite (xManager).apk
04-13 08:05:09.429  5825  5825 W PackageManager: 	at android.content.pm.PackageParser.parseApkLiteInner(PackageParser.java:1512)
04-13 08:05:09.429  5825  5825 W PackageManager: 	at android.content.pm.PackageParser.parseApkLite(PackageParser.java:1483)
04-13 08:05:09.429  5825  5825 W PackageManager: 	at android.content.pm.PackageParser.parseMonolithicPackageLite(PackageParser.java:989)
04-13 08:05:09.429  5825  5825 W PackageManager: 	at android.content.pm.PackageParser.parseMonolithicPackage(PackageParser.java:1198)
04-13 08:05:09.429  5825  5825 W PackageManager: 	at android.content.pm.PackageParser.parsePackage(PackageParser.java:1112)
04-13 08:05:09.429  5825  5825 W PackageManager: 	at android.content.pm.PackageManager.getPackageArchiveInfo(PackageManager.java:8074)
04-13 08:05:09.429  5825  5825 W PackageManager: 	at android.content.pm.PackageManager.getPackageArchiveInfo(PackageManager.java:8051)
04-13 08:05:09.429  5825  5825 W PackageManager: 	at com.xc3fff0e.xmanager.MainActivity._Downgrade_Checker(MainActivity.java:14952)
04-13 08:05:09.429  5825  5825 W PackageManager: 	at com.xc3fff0e.xmanager.MainActivity$127$1$1$1.run(MainActivity.java:15768)
04-13 08:05:09.429  5825  5825 W PackageManager: 	at android.os.Handler.handleCallback(Handler.java:942)
04-13 08:05:09.429  5825  5825 W PackageManager: 	at android.os.Handler.dispatchMessage(Handler.java:99)
04-13 08:05:09.429  5825  5825 W PackageManager: 	at android.os.Looper.loopOnce(Looper.java:201)
04-13 08:05:09.429  5825  5825 W PackageManager: 	at android.os.Looper.loop(Looper.java:288)
04-13 08:05:09.429  5825  5825 W PackageManager: 	at android.app.ActivityThread.main(ActivityThread.java:7924)
04-13 08:05:09.429  5825  5825 W PackageManager: 	at java.lang.reflect.Method.invoke(Native Method)
04-13 08:05:09.429  5825  5825 W PackageManager: 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
04-13 08:05:09.429  5825  5825 W PackageManager: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:936)

```

> Do you have to download an APK for the spam to start happening? If not, does something specific trigger this?

No, just open the app.

> if so, can you confirm with a File manager, adb or whatever that a file exists (or doesn't exist) at the path(s) being spammed in the log?

The folder exists, but no file inside. Can I download the correct file anywhere to just put it inside and see what's happening? 


I have some experience in Android programming and will take a closer look at the project and possibly debug it when I have time.

**Comment by dchristl** [comment]
I've looked through the code a bit. Unfortunately, I couldn't compile it myself because the project seems to rely on a very old Gradle version, and updating it would require significant changes. However, with the help of the stack trace, I got closer to understanding the issue. 

The main issue here is that the checked-in code (Tag 5.4) doesn't match the generated APK (line numbers doesn't fit). How is the build process set up? Where are the correct sources located?

Nevertheless, based on the code and the stack traces, I made some progress. In the UI thread, a method called _Downgrade_Checker is called every **75 ms**. This method scans all theoretically downloaded APKs (6 in total in the specified folder) and sets the version numbers for the UI. It's puzzling why this happens so often and why, for example, the Wave is still included, even though it's no longer downloadable. This method throws the error for all non-existent APKs, resulting in 6 exceptions every 75 ms.

But this doesn't seem to be a Lineage-specific issue; rather, it appears to be a general one since I didn't see any functionality that automatically downloads all APKs. Everyone has this problem because Wave is no longer downloadable, and usually, the Amoled and stock versions probably aren't downloaded in parallel either. Moreover, if you use the mirror link, it ends up in a different folder anyway.

Now I'm stuck because the source code doesn't match the compiled output. In the source code, every access to the APKs is caught with a catch exception, but apparently not in the apk version.

The code probably needs some improvement, such as performing a file existence check beforehand (which is also checked during download) and removing the polling every 75ms.

**Comment by energ0n** [comment]
> Yes, because download works.
> 
> It's always the same output over and over again and stops only on force closing the app.

Thanks for all the info again!

After a second look at everything I now see what's going on. I originally thought this was a storage access issue related to additional Lineage privacy guard controls (which I then learned have been removed since ~v17) and assumed the app was unable to save and/or install the downloaded APK.

A proper logcat with xManager isolated and I now see this same "log spam" with download locations constantly being checked. Not ideal behavior.. to put it mildly. 

I'll ping @xC3FFF0E and maybe this can be addressed in a future builds.
